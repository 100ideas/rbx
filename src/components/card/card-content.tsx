import classNames from "classNames";
import * as React from "react";

import { forwardRefAs, Generic, HelpersProps } from "src/base";

export type CardContentProps = HelpersProps;

export const CardContent = forwardRefAs<CardContentProps, "div">(
  ({ className, ...rest }, ref) => (
    <Generic
      className={classNames("card-content", className)}
      ref={ref}
      {...rest}
    />
  ),
  { as: "div" },
);
