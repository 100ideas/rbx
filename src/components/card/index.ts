import "./card.sass";

export { default } from "./Card";
