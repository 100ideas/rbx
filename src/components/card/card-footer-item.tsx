import { cx } from "emotion";
import React from "react";

import { asExoticComponent } from "@/components/exotic";
import { ModifierProps, modify } from "@/modifiers";

export type CardFooterItemProps = ModifierProps;

export const CardFooterItem = asExoticComponent<CardFooterItemProps, "div">(
  (props, ref) => {
    const { as, ...rest } = modify(props);
    rest.className = cx("card-footer-item", rest.className);
    return React.createElement(as!, { ref, ...rest });
  },
  "div",
);
