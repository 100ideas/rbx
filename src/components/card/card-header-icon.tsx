import { cx } from "emotion";
import React from "react";

import { asExoticComponent } from "@/components/exotic";
import { ModifierProps, modify } from "@/modifiers";

export type CardHeaderIconProps = ModifierProps;

export const CardHeaderIcon = asExoticComponent<CardHeaderIconProps, "div">(
  (props, ref) => {
    const { as, ...rest } = modify(props);
    rest.className = cx("card-header-icon", rest.className);
    return React.createElement(as!, { ref, ...rest });
  },
  "div",
);
