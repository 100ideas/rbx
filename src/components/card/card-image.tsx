import { cx } from "emotion";
import React from "react";

import { Element } from "@/components/element";
import { Image } from "@/components/image";
import { ImageProps } from "@/components/image/image";

export type CardImageModifierProps = Partial<{
  className: string;
}>;

export type CardImageProps = ImageProps & CardImageModifierProps;

export const CardImage = React.forwardRef<HTMLElement, CardImageProps>(
  ({ className, ...rest }, ref) => (
    <Element className={cx("card-image", className)}>
      <Image ref={ref} {...rest} />
    </Element>
  ),
);
