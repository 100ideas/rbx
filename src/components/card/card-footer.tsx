import { cx } from "emotion";
import React from "react";

import { asExoticComponent } from "@/components/exotic";
import { ModifierProps, transformModifiers } from "@/modifiers";
import { CardFooterItem } from "./card-footer-item";

export type CardFooterProps = ModifierProps;

export const CardFooter = Object.assign(
  asExoticComponent<CardFooterProps, "div">((props, ref) => {
    const { as, ...rest } = transformModifiers(props);
    rest.className = cx("card-footer", rest.className);
    return React.createElement(as!, { ref, ...rest });
  }, "div"),
  { Item: CardFooterItem },
);
