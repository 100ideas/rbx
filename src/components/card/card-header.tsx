import { cx } from "emotion";
import React from "react";

import { asExoticComponent } from "@/components/exotic";
import { ModifierProps, modify } from "@/modifiers";
import { CardHeaderIcon } from "./card-header-icon";
import { CardHeaderTitle } from "./card-header-title";

export type CardHeaderProps = ModifierProps;

export const CardHeader = Object.assign(
  asExoticComponent<CardHeaderProps, "div">((props, ref) => {
    const { as, ...rest } = modify(props);
    rest.className = cx("card-header", rest.className);
    return React.createElement(as!, { ref, ...rest });
  }, "div"),
  {
    Icon: CardHeaderIcon,
    Title: CardHeaderTitle,
  },
);
