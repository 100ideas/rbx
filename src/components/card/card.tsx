import { cx } from "emotion";
import React from "react";

import { Element, renderAsExoticComponent } from "@/components/element";
import { ModifierProps } from "@/modifiers";
import { CardContent } from "./card-content";
import { CardFooter } from "./card-footer";
import { CardHeader } from "./card-header";
import { CardImage } from "./card-image";

export type CardModifierProps = Partial<{
  children?: React.ReactNode;
}>;

export type CardProps = ModifierProps & CardModifierProps;

export const Card = Object.assign(
  renderAsExoticComponent<CardProps, "div">(
    ({ className, children, ...props }, ref) => (
      <Element ref={ref} className={cx("card", className)} {...props}>
        {children}
      </Element>
    ),
    "div",
  ),
  {
    Content: CardContent,
    Footer: CardFooter,
    Header: CardHeader,
    Image: CardImage,
  },
);
Card.defaultProps = Object.assign({ children: null }, Card.defaultProps);
