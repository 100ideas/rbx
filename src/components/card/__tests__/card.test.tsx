import Enzyme from "enzyme";
import React from "react";

import { Card } from "../card";
import { CardContent } from "../card-content";
import { CardFooter } from "../card-footer";
import { CardHeader } from "../card-header";
import { CardImage } from "../card-image";

import { hasProperties, testGenericPropTypes } from "@/__tests__/testing";

describe("Card component", () => {
  hasProperties(Card, {
    Content: CardContent,
    Footer: CardFooter,
    Header: CardHeader,
    Image: CardImage,
    defaultProps: { as: "div" },
  });

  it("should render as the default element", () => {
    const wrapper = Enzyme.shallow(<Card />);
    expect(wrapper.is("div")).toBe(true);
  });

  it("should render as a custom component", () => {
    const as = "span";
    const wrapper = Enzyme.shallow(<Card as={as} />);
    expect(wrapper.is(as)).toBe(true);
  });

  it("should forward ref", () => {
    const ref = React.createRef<HTMLDivElement>();
    // Enzyme owns outer ref: https://github.com/airbnb/enzyme/issues/1852
    const wrapper = Enzyme.mount(
      <div>
        <Card ref={ref} />
      </div>,
    );
    try {
      expect(ref.current).toBe(wrapper.find(".card").instance());
    } finally {
      wrapper.unmount();
    }
  });

  it("should have bulma className", () => {
    const wrapper = Enzyme.shallow(<Card />);
    expect(wrapper.hasClass("card")).toBe(true);
  });

  it("should preserve custom className", () => {
    const className = "foo";
    const wrapper = Enzyme.shallow(<Card className={className} />);
    expect(wrapper.hasClass(className)).toBe(true);
  });

  describe("propTypes", () => {
    const { propTypes } = Card;
    testGenericPropTypes(propTypes);
  });
});
