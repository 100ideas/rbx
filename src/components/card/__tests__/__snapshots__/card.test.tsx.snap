// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component should have card classname 1`] = `
<div
  className="card"
>
  Card Content
</div>
`;

exports[`Card component should have card-content classname 1`] = `
<div
  className="card-content"
>
  Content
</div>
`;

exports[`Card component should have card-footer's classname 1`] = `
<div
  className="card"
>
  <div
    className="card-footer"
  >
    <div
      className="card-footer-item"
    >
      Yes
    </div>
    <div
      className="card-footer-item"
    >
      No
    </div>
  </div>
</div>
`;

exports[`Card component should have card-header's classname 1`] = `
<div
  className="card"
>
  <div
    className="card-header"
  >
    <div
      className="card-header-title"
    >
      Title
    </div>
    <div
      className="card-header-icon"
    >
      <i
        className="icon"
      />
    </div>
  </div>
</div>
`;

exports[`Card component should have card-image classname 1`] = `
<div
  className="card-image"
>
  <figure
    className="image is-4by3"
  >
    <img
      src="http://bulma.io/images/placeholders/1280x960.png"
    />
  </figure>
</div>
`;
